lib/hanami/config/actions.rb

```diff
@@ -99,6 +99,7 @@ def initialize(*, **options)
         super()
 
         @base_config = Hanami::Action.config.dup
+        @base_config.root_directory = options[:root_directory] if options.key?(:root_directory)
```

Comment:
@timriley I started applying this change, but I stopped because of the feedback from the tests. 🙈 

1. Hundreds of unit test failures because each time we initialize `Hanami::Config::Actions` in a test, it would require a `Hanami::Config` instance.
2. From a design perspective, passing `self` to a dependency is a red-flag of cyclic dependency.
3. From a perf perspective, we're passing a large object (`Hanami::Config`) to the other `Hanami::Config::Actions`.

---

🤔 So I stopped to ask myself: "Why do we need `root_directory` in actions"?

It's only used indirectly by a feature `Hanami::Action::Response#send_file`.
This method allows to send a file to a browser.

For security reasons, we need to specify a base directory to prevent attackers from navigating the server file system.
This base directory is `public_directory`, calculated from `root_directory.join("public")`.

---

👉 Proposal: keep this feature, but implement it in `hanami`. So we won't need to propagate `Hanami::Config#root_directory` to `Hanami::Config::Actions`.

Reply:
@jodosha Thanks for sharing your thinking. Just to confirm: you're proposing here that we might implement the entire `Response#send_file` feature in hanami instead of hanami-controller?

If this is the case, I don't feel particularly excited by that approach, because `#send_file` is a feature that in every other way fits better within hanami-controller: it's an HTTP-specific operation, is still useful independently of the full Hanami framework, and for those reasons, would be better maintained alongside the other features of hanami-controller.

I also feel like the result could be messier than our idea here of simply passing `self` to the nested `Hanami::Config::Actions`: we'd need to implement a new extension to `Response` here where none existed before, which diffuses the responsibilities of hanami-controller.

---

All of that said, I do agree with you that passing the app config object to a nested config could be a smell.

However, if I think about our post-2.0 future world—where we expose public, framework-provided integration hooks for extension gems—one of the things I expect these gems will need to do is to be able to adapt and configure themselves based on the app's base configuration, which is the same thing we're looking to do here.

The one distinction at this point is that we currently control both sides of the equation: we're writing both the "extension gem" (which is the `Hanami::Config::Actions` wrapper around the base `Hanami::Action` config) as well as the framework-level code that is hosting it. And because these _look_ like they're long-term parts of a single codebase, it feels like the arrangement should be different, for all the regular principles that apply to good internal separation of concerns, etc.

---

I wonder if there might be other ways we could consider solving this... instead of passing the entire app config to the nested config object at initialise-time...

Could we instead pass it as part of the `finalize!` step? So we'd end up code in the nested config class looking something like this:

```ruby
      def finalize!(app_config)
        # <snip other code>

        @base_config.root_directory = app_config.root
      end
```

That's extremely similar to the proposal I shared above, but without the benefit of a decreased coupling between the two objects: now the nested config no longer has the app config for the entirety of its lifecycle; instead it receives it for a one time opportunity to make sure its nested config can be appropriately synced up with any relevant configs at the app level.

What do you think about an adjustment like this?

I toyed around with it quickly and here it is with tests still passing: https://github.com/hanami/hanami/compare/feature/serve-static-assets...feature/serve-static-assets--app-config-when-finalizing

Note: that branch is just demonstrating the change of passing the `app_config` argument to `finalize!` — I haven't done anything _with_ the `app_config` yet. And to simplify one set of test changes, I did have to remove customisation of the CSP based on a `assets_server_url:` argument, but looking through our actual runtime (non-test) code, we don't actually provide that argument at all! So this is probably a useful elimination of dead code.

---

Another possible option here is just having our nested config classes simply reaching out directly to `Hanami.app.config` for things they need. This is gross, yes, and this would be an acknowledged act of "kicking the can down the road" in terms of feeling our way towards what a clean extension API for config might look like. To be honest, given this is all entirely internal implementation detail right now, with none of these changes having any bearing on our public API, I feel comfortable with quite a range of approaches here (shocking, I know!).

And here's why: at the end of the day, changes to these config internals are all "two-way doors" — we preserve the ability to change course in the future without breaking any compatibility promises to our users.

On the other hand, by taking what is on its own a fairly useful part of the hanami-controller (`Response#send_file`) and moving it into hanami, we're actually making a potentially breaking change to standalone use of hanami-controller, which is already released as 2.0. I'd like to see us avoid this. Aside from the API breakage, I like the principle of our gems being as useful as possible in standalone mode, and I think it's a better call to honour that here.

Reply:
@timriley Wow, thanks for getting deep with your answers.

To summarize:

* Let's keep `send_file` in `hanami-controller`
* Let's pass config to `#finalize!`
* Later internal architecture to hook into the booting process of Hanami.

I cherry-picked your [commit](https://github.com/hanami/hanami/commit/a6289df4ad98e12381ade7d729cf2e8b7ee1ab94) and pushed it here. Thanks for that too!